src/header/common/range.rs

```diff
@@ -0,0 +1,199 @@
+use std::fmt::{self, Display};
+use std::str::FromStr;
+
+use header::{Header, HeaderFormat, RangeUnit};
+use header::parsing;
+
+/// The `Range` header defined by HTTP/1.0.
+///
+/// > The "Range" header field on a GET request modifies the method
+/// > semantics to request transfer of only one or more subranges of the
+/// > selected representation data, rather than the entire selected
+/// > representation data.
+///
+/// > Range = byte-ranges-specifier / other-ranges-specifier
+/// > other-ranges-specifier = other-range-unit "=" other-range-set
+/// > other-range-set = 1*VCHAR
+///
+/// # Example values
+/// * `bytes=1000-`
+/// * `bytes=-2000`
+/// * `bytes=0-1,30-40`
+/// * `custom_unit=0-123`
+///
+/// Spec: https://tools.ietf.org/html/rfc7233#section-3.1
+#[derive(PartialEq, Clone, Debug)]
+pub struct Range {
+    /// Unit of the Range i.e. bytes
+    pub unit: RangeUnit,
+    /// Set of ranges as defined in the HTTP spec
+    pub ranges: Vec<RangeSpec>,
+}
+
+/// Each 'Range' header can contain one or more RangeSpecs.
+/// Each RangeSpec defines a range of units to fetch
+#[derive(PartialEq, Clone, Debug)]
+pub enum RangeSpec {
+    /// Get all bytes between x and y ("x-y")
+    FromTo(u64, u64),
+    /// Get all bytes starting from x ("x-")
+    AllFrom(u64),
+    /// Get last x bytes ("-x")
+    Last(u64)
+}
+
+impl Range {
+    /// Get the most common byte range header ("bytes=from-to")
+    pub fn bytes(from: u64, to: u64) -> Range {
+        Range {
+            unit: RangeUnit::Bytes,
+            ranges: vec![RangeSpec::FromTo(from, to)],
+        }
+    }
+}
+
+
+impl fmt::Display for RangeSpec {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        match *self {
+            RangeSpec::FromTo(from, to) => write!(fmt, "{}-{}", from, to),
+            RangeSpec::Last(pos) => write!(fmt, "-{}", pos),
+            RangeSpec::AllFrom(pos) => write!(fmt, "{}-", pos),
+        }
+    }
+}
+
+
+impl fmt::Display for Range {
+    fn fmt(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
+        try!(write!(fmt, "{}=", self.unit));
+
+        let mut ranges = self.ranges.iter();
+        match ranges.next() {
+            Some(range) => try!(write!(fmt, "{}", range)),
+            None => return Ok(())
+        }
+        for range in ranges {
+            try!(write!(fmt, ",{}", range));
+        }
+        Ok(())
+    }
+}
+
+impl FromStr for Range {
+    type Err = ();
+
+    fn from_str(s: &str) -> Result<Range, ()> {
+        let mut iter = s.split("=");
+
+        let (unit, ranges) = match (iter.next(), iter.next()) {
+            (Some(unit), Some(ranges)) => (unit, ranges),
+            _ => return Err(())
+        };
+
+        if let Ok(unit) = RangeUnit::from_str(unit) {
+            let mut valid_ranges: Vec<RangeSpec> = vec![];
+            for range in ranges.split(",") {
```

Comment:
There can be optional whitespace between the values and empty values like `bytes=100-200, , , 300-400, -900`.

The ABNF is:
```
     byte-ranges-specifier = bytes-unit "=" byte-range-set
     byte-range-set  = 1#( byte-range-spec / suffix-byte-range-spec )
     byte-range-spec = first-byte-pos "-" [ last-byte-pos ]
     first-byte-pos  = 1*DIGIT
     last-byte-pos   = 1*DIGIT
```

The definition for `1#` is given in https://tools.ietf.org/html/rfc7230#section-7 optional whitespce is allowed but not required. You can probably use `from_one_comma_delimited()` here.